### ISSUE TITLE
colorspace conversion seems wrong

### ISSUE THREAD

**Issue by unic0rn** [mpv-player/mpv#1557]
windows, radeon 9550 (yeah, that one on agp), video output - both direct3d_shaders and direct3d. 

checked on several movies, compared with some stills from the web and with mplayer using opengl output (opengl in mpv sadly doesn't work on 9550 for me) - the balance is very slightly wrong, for example large green areas (grass and such stuff) seem to have a very little less green component, but at the same time faces get a slightly yellow tint. you would have to compare it directly with mplayer using opengl output, frame to frame, to see the difference, but it's there - and mplayer's output exactly matches stills from the web (and i somehow doubt they were all done using mplayer with that output), while mpv's output, well. i wouldn't really be sure what is wrong and what is not, but that yellowish tint on faces just seems wrong. and yes, i've calibrated my monitor properly.

**Comment by unic0rn** [comment]
i've checked if it's a matter of colorspace matrix and it's not. what's more interesting is the fact that mplayer with video output set to direct3d generates same picture as mpv (kinda obvious actually, i know). makes me wonder if there's a difference between colorspace conversion results using opengl and direct3d in mpv.

i'm just wondering where's the error, because two separate results cannot be correct, given the exact colorspace matrix that should be used.

word of explanation - i didn't use direct3d output in mplayer earlier because on previous gpu it was just wrong, and opengl was working fine so i've sticked to that. and now i'm just confused - is there similar difference in opengl and direct3d output in mpv? if so, why? and regardless, which output is more accurate? i would prefer to use mpv, but above that i would prefer to use the player that displays correct picture - and opengl in mpv on this gpu is a no go.

**Comment by wm4** [comment]
Does direct3d_shaders actually use shaders for you? Shaders are automatically disabled if they can't be created.

Plain direct3d uses the GPU's builtin method for scaling and color conversion, which seems often to lead to broken results.

Why would opengl not work?

**Comment by unic0rn** [comment]
it says it does use shaders.

[vo/direct3d_shaders] Accepted rendering methods for format='yuv420p': StretchRect=0x323
15659, Texture=0, Texture+Shader=0x32.
[vo/direct3d_shaders] Using 3D rendering.
[vo/direct3d_shaders] Using YUV shaders.
[vo/direct3d_shaders] resize_d3d 1280x720 called.
[vo/direct3d_shaders] create_d3d_surfaces called.

the gpu supports shader model 2.0, so it is dx9 capable. opengl doesn't work because it supports only opengl 2.0 and 2.1 is required at minimum. mplayer with opengl output uses shaders as well.

i've tried changing the colormatrix with direct3d_shaders output and there are subtle differences, so it works. the thing is, the output is different than what mplayer generates with opengl - and seems the same that mplayer generates with direct3d (and i'm not sure if mplayer uses shaders for that). so it seems that plain direct3d in mplayer gives the same results as plain direct3d in mpv and direct3d_shaders in mpv - so everything seems to be working perfectly, with the exception that opengl in mplayer gives different output and while the difference is subtle, it's there - and it is bigger - as far as i've noticed - than what you get by changing the colorspace matrix (i've tried changing the colorspace in mplayer as well, i've also checked software colorspace conversion in mplayer - no difference from shaders).

**Comment by wm4** [comment]
What if you try ``--vf=format=rgba``?

>mplayer generates with direct3d (and i'm not sure if mplayer uses shaders for that).

No.

**Comment by unic0rn** [comment]
black picture. log: http://wklej.org/hash/b2c8cf9f749/txt/

**Comment by unic0rn** [comment]
sorry for the quiet option in config. with --quiet=no it looks like all the frames are being dropped.

**Comment by unic0rn** [comment]
and with --quiet=no --framedrop=no no framedrops, still black picture, and just like with framedrop, A-V constantly goes south.

(can't edit old comments in ff17, and that's the last browser with working jit on non-sse2 cpu)

**Comment by unic0rn** [comment]
and to make it clear what kind of difference i'm talking about, two frames, each shot with mplayer using opengl output and mpv using direct3d_shaders:

#1 mplayer http://wstaw.org/m/2015/02/04/mplayer1.png
#1 mpv http://wstaw.org/m/2015/02/04/mpv1.png
#2 mplayer http://wstaw.org/m/2015/02/04/mplayer2.png
#2 mpv http://wstaw.org/m/2015/02/04/mpv2.png

on 'normal movies' the difference is obviously less visible because of smaller saturation, but it's there - and it's even hard to guess which output is the correct one. on tangled it seems like mplayer via opengl gets it right, but i wouldn't bet on it. besides, as i've said, mplayer via direct3d (no shaders as you've said) looks the same as mpv via direct3d - with or without shaders. and shaders work i think.

just in case, mpv log - exactly the same settings as used for the screenshots above, the only option added was -v: http://wklej.org/hash/31d8a1d6b96/txt/

**Comment by wm4** [comment]
Can you make a screenshot with mpv's screenshot function of the exact same frame? It doesn't involve the VO, and should give a relatively correct result.

**Comment by unic0rn** [comment]
yeah, that comment that only -v was added, was a little bit obscure. i did use --vf screenshot for both mplayer and mpv screenshots posted above.

**Comment by unic0rn** [comment]
also, for me those screenshots look exactly like what was on the screen in mpv's window. how does --vf screenshot not involve the VO? the colorspace conversion is done by the VO so it's obviously the output of the VO that was captured... or is it getting decoded frame and doing colorspace conversion in software for the purpose of such screenshot? if so, the results are exactly the same.

**Comment by wm4** [comment]
> i did use --vf screenshot for both mplayer and mpv screenshots posted above.

Then both use libswscale, and it's probably a difference in the colorspace passed to libswscale. I bet this is just a MPlayer bug not handling bt.709 correctly.

>how does --vf screenshot not involve the VO? the colorspace conversion is done by the VO so it's obviously the output of the VO that was captured...

No, it uses libswscale and the VO is not involved. There's also a screenshot mode that captures window output (``screenshot window``), but you probably didn't use it.

**Comment by unic0rn** [comment]
as i understand you, libswscale is used by the --vf screenshot, because i cannot find any info about libswscale being used in regular playback in verbose log from neither mplayer nor mpv - and colorspace conversion is done by the VO in both cases.
and in both cases it looks exactly as those screenshots, which were made using --vf screenshot (so i guess using libswscale).

so either something doesn't add up or i don't understand something.

**Comment by unic0rn** [comment]
i just did a simple test - it doesn't prove that everything is ok, but it does prove - or so i think - that there's no major screwup somewhere.

i've converted a gamma 2.2 chart from .png to .mkv using ffmpeg, with -c:v ayuv to prevent chart corruption.

mplayer with opengl output displays it fine, it looks exactly like the original. mpv tries to feed direct3d_shaders with bgra, resulting in black screen. with --vf format=yuv444p --colormatrix BT.601 it displays fine (mplayer defaults to 601, at least at this file it seems). that means that the colorspace conversion worked fine in this case.

ok. another test done. mea culpa, i haven't noticed it earlier (not enough frame-to-frame comparisons, and mplayer doesn't say a word about colorspace with -v).

basically, mplayer defaults to BT.601. what does keep me wondering though, is the fact that after adjusting that there are still differences. i did two tests, in one i've enforced BT.601 on mpv, in another i've set mplayer to use BT.709. in both cases the test frame had exactly the same (to my tired eyes at least) colour balance, but the difference was in saturation. direct3d_shaders produces visibly more saturated output than opengl in mplayer.

the question remains why and which method is the correct one.

should i guess, i would say it may have something to do with levels conversion. mplayer's manual mentions brightness levels only, at least in case of opengl. perhaps someone forgot to convert chroma. what doesn't fit such scenario for me though is the fact that i suspect it should screw up the colours - and with correct colorspace set, everything seems to look fine, it's only the saturation that's lower. should it be? is direct3d output producing oversaturated picture?

**Comment by wm4** [comment]
mpv should handle colorspace tags pretty well with the opengl output, and even the direct3d_shaders output (though with plain direct3d output it depends on the GPU drivers). Also this looks like HD material, which normally uses bt.709. Also, greens look less saturated in bt.709. So I'd say all signs point to mpv being correct here.

**Comment by haasn** [comment]
It's hard to tell when opening these kinds of files in Firefox or other color managed viewers, since then they're also suddenly interpreted as sRGB and whatnot.

Here is a version of the files with colorspace tags appended:

mplayer1.png: ![](http://i.srsfckn.biz/a97.png)
mpv1.png: ![](http://i.srsfckn.biz/918.png)
mplayer2.png: ![](http://i.srsfckn.biz/54.png)
mpv2.png: ![](http://i.srsfckn.biz/357.png)

It's hard to tell, but I think based on the brightness of the stuff in the background, the ‘mpv’ output is more correct, even though the skin tone is much uglier. It might just be a mis-tagged source file, though. Who knows?

All I can say is that if you want colors that are definitely correct, you should *only* use -vo opengl, as everything else has basically zero guarantee of even remotely resembling the source.

**Comment by unic0rn** [comment]
well, those files were generated by mplayer and mpv, so as for incorrect tags...
also, they look the same for me in the browser as the source does in the player, if i would notice any difference i wouldn't post them.

and yeah, the skin tone is uglier with BT.709, but it seems that's how it should be.

@wm4 : there's one more thing to this as i've mentioned. you are right that mpv handles BT.709 correctly, but there's still a difference in how it's handled in mpv and in mplayer with opengl. the picture is visibly more saturated in mpv. wether it's because of wrong chroma conversion in mplayer or something else, no idea. i don't think missing chroma conversion would end up with exactly same colours only less saturated, but i'm no expert. ~~but it's certainly not just messed up BT.709 in mplayer, because when both are forced to BT.601, the difference is the same (just colours in both players obviously different thanks to BT.601) - less saturated picture in mplayer. sadly, no screenshots - i've tried and mplayer with -vf screenshot produces screens using BT.601 instead, which is kinda obvious since it's a parameter for VO which isn't used for screens, and i have no idea how - if it's possible - to tell libswscale in mplayer to use BT.709 (and i didn't want third party software to get involved in taking those screens).~~

ok, figured. these are the colorspace options for mplayer's opengl output:
0 - Use the formula used normally by MPlayer (default).
1 - Use ITU-R BT.601 color space.
2 - Use ITU-R BT.709 color space.
3 - Use SMPTE-240M color space.

i didn't test 3 because i'm already getting grey hair out of this.
1 - so BT.601 - produces same colours as mpv with BT.601 enforced, just less saturated
2 - so BT.709 - produces same colours as mpv by default (so using BT.709), just less saturated
0 - so BT.601 (one would expect) - produces exactly same picture as mpv with BT.601!

so basically opengl output in mplayer generates proper picture with default setting only, and that default is BT.601. trying to choose specifically either BT.601 or BT.709 ends up in undersaturated picture.

sorry for the mess and thanks for your help. i guess i'll throw it at mplayer's team then and i'll stick to mpv.

**Comment by unic0rn** [comment]
so yeah, the screenshots i've posted earlier, in mplayer's case are BT.601 - and that's proper BT.601 at that, exactly the same what you would get from mpv --colormatrix BT.601.

**Comment by wm4** [comment]
It's possible that MPlayer has the wrong colormatrices as well. (I think we fixed or rewrote that code in mplayer2 times).

**Comment by pigoz** [comment]
Yeah mplayer had wrong colormatrixes. uau rewrote all that stuff in mplayer2 so that they are generated instead of hard coded.

**Comment by haasn** [comment]
For the sake of documentation, I came across a raw BD remux of this scene and decided to create a more or less correct “reference” image to show that's correct.

The methodology used was to configure mpv with target-prim=bt709:target-trc=gamma22 and attach a bt709/gamma 2.2 ICC profile to the PNG file.

![](https://0x0.st/Pl.png)

**Comment by unic0rn** [comment]
"source"?

with all the respect to your effort, compare your images with mine. agreed, mine has a few pixels on the sides cropped out for unknown reasons, so it's probably a crappy rip, but at least it doesn't have all the background blurred into oblivion.

something is very wrong with your "source".

**Comment by lachs0r** [comment]
Nothing is wrong with that source. It’s straight from the BDs.
If anything, yours is oversharpened (see ringing artifacts).

**Comment by haasn** [comment]
Mine is a straight-up BD remux. This is the H.264 bitstream as it came directly from the producer of the film, with no filtering or postprocessing applied other than the color conversions by mpv.

Note that due to chroma subsampling, two results are never going to be bit-exact unless they use the exact same chroma interpolation algorithm - it's possible that my chroma interpolator was slightly less sharp than yours, but it's doubtful that it would make a huge difference here.

Either way, it won't impact the colors.

**Comment by unic0rn** [comment]
chroma subsampling? slightly less sharp?

you're either joking or blind. i'm not sure what kind of postprocessing that BD went through, but it has TONS of details blurred out. look at the background on the left:

http://wstaw.org/m/2015/02/04/mpv2.png

and compare it with your screenshot. yours looks like it has some very heavy depth of field filter. not sure if it's on purpose by disney - in theory it may be used to add a depth effect to a 2D release of a movie that was 3D iirc, and i'm not sure if such thing could be done in 2D only - but damn, it just looks wrong.

if you say it's like that on the original BDs - fine, but don't blame it on oversharpening and such stuff, because it's not my release that looks wrong. if sharpening would produce the amount of details my copy has that official BDs are lacking, some people would pay billions for it.

**Comment by lachs0r** [comment]
Well, looks like that’s just what Disney did there. Anyway, the point of this issue was on colors, not differences between releases of the same movie.
